packages/wdio-browserstack-service/src/util.ts

```diff
@@ -1359,3 +1482,66 @@ export const isValidCapsForHealing = (caps: { [key: string]: Options.Testrunner
     // Check if there are any capabilities and if at least one has a browser name
     return capValues.length > 0 && capValues.some(hasBrowserName)
 }
+
+type PollingResult = {
+    data: any;
+    headers: Record<string, any>;
+    message?: string; // Optional message for timeout cases
+  };
+
+async function pollApi(
+    url: string,
+    params: Record<string, any>,
+    headers: Record<string, string>,
+    upperLimit: number,
+    startTime = Date.now()
+): Promise<PollingResult> {
+    try {
+        params.timestamp = Date.now() / 1000
+        console.log(`current timestamp ${params.timestamp}`)
+        const response = await axios.get(url, { params, headers })
+
+        // If the request succeeds (non-404), return the result
+        return {
+            data: response.data,
+            headers: response.headers,
+            message: 'Polling succeeded.',
+        }
+    } catch (error) {
+        if (axios.isAxiosError(error) && error.response?.status === 404) {
+            const nextPollTime = parseInt(error.response.headers.next_poll_time, 10) * 1000
+            console.log(`timeInMillis ${nextPollTime}`)
+            if (isNaN(nextPollTime)) {
+                console.warn('Invalid or missing `nextPollTime` header. Stopping polling.')
+                return {
+                    data: null,
+                    headers: error.response.headers,
+                    message: 'Invalid timeInMillis header value. Polling stopped.',
+                }
+            }
+
+            const elapsedTime = nextPollTime - Date.now()
+            console.log(`elapsedTime ${elapsedTime} timeInMillis ${nextPollTime} upperLimit ${upperLimit}`)
+            // Stop polling if the upper time limit is reached
+            if (nextPollTime > upperLimit) {
+                console.log('Polling stopped due to upper time limit.')
+                return {
+                    data: null,
+                    headers: error.response.headers,
+                    message: 'Polling stopped due to upper time limit.',
+                }
+            }
+
+            console.log(`Polling again in ${elapsedTime}ms with params:`, params)
+
+            // Wait for the specified time and poll again
+            await new Promise(resolve => setTimeout(resolve, elapsedTime))
+
+            // Recursive call
+            return pollApi(url, params, headers, upperLimit, startTime)
+        }
+        console.error('Unexpected error occurred:')
+        return { data: {}, headers: {}, message: 'Unexpected error occurred.' }
```

Comment:
messages should be coming from backend right?

Reply:
Handled